### PR TITLE
Convert `_insert_graph_receipts_txn` to `simple_upsert`

### PR DESCRIPTION
**Pull request by realtyem**
Which should cut down on Postgres log spam like:
```log
2023-09-11 05:37:47.811 CDT [13934] ERROR:  could not serialize access due to concurrent update
2023-09-11 05:37:47.811 CDT [13934] STATEMENT:  INSERT INTO receipts_graph (room_id, receipt_type, user_id, thread_id, event_ids, data) VALUES ('!room_1:computer.surgery', 'm.read', '@some_redacted_user:gnome.org', 'main', '["$hCBJpUJwHjjG6ztFwHrDPtqHFyMQ8eWMOCRUhmO3UBs"]', '{"thread_id":"main","ts":1694427550626}') ON CONFLICT (room_id, receipt_type, user_id, thread_id)  DO UPDATE SET event_ids=EXCLUDED.event_ids, data=EXCLUDED.data
2023-09-11 05:52:03.258 CDT [13934] ERROR:  could not serialize access due to concurrent update
2023-09-11 05:52:03.258 CDT [13934] STATEMENT:  INSERT INTO receipts_graph (room_id, receipt_type, user_id, thread_id, event_ids, data) VALUES ('!alCakyySsFIAVfZLDL:matrix.org', 'm.read', '@some_other_redacted_user:personal_homeserver.uk', 'main', '["$ekmtqEQ5PFNHFg49-zZamgw6LDNcwqYMh0dyGSQ3A2E"]', '{"thread_id":"main","ts":1694429513585}') ON CONFLICT (room_id, receipt_type, user_id, thread_id)  DO UPDATE SET event_ids=EXCLUDED.event_ids, data=EXCLUDED.data
2023-09-11 05:57:43.559 CDT [14238] ERROR:  could not serialize access due to concurrent update
2023-09-11 05:57:43.559 CDT [14238] STATEMENT:  INSERT INTO receipts_graph (room_id, receipt_type, user_id, thread_id, event_ids, data) VALUES ('!no_idea_room_id:matrix.org', 'm.read', '@another_random_user_id:other_homeserver.pt', 'main', '["$nBSxdTs9h73YOLm40A-ggrfSOvv7KL-iJHOaRzOjMnY"]', '{"thread_id":"main","ts":1694429518348}') ON CONFLICT (room_id, receipt_type, user_id, thread_id)  DO UPDATE SET event_ids=EXCLUDED.event_ids, data=EXCLUDED.data
```

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: Jason Little <realtyem@gmail.com>

### CODE REVIEW

**Review by erikjohnston — CHANGES_REQUESTED** [review]
It looks like `_insert_graph_receipt_txn` is basically just an upsert, which we can do as an "autocommit". I'd suggest we change  `_insert_graph_receipt_txn` to use `await simple_upsert(..)` instead so we get that for free. 

This safes having to reason about if its "safe" or not.

**Comment by realtyem** [comment]
> It looks like `_insert_graph_receipt_txn` is basically just an upsert, which we can do as an "autocommit". I'd suggest we change `_insert_graph_receipt_txn` to use `await simple_upsert(..)` instead so we get that for free.
> 
> This safes having to reason about if its "safe" or not.

Ooooooooo, neat idea. I did not know you could do that. Do we ever actually read this table?

**Comment by clokep** [comment]
> Do we ever actually read this table?

No and I debated about removing it. It is somewhat there for "correctness" but we never implemented that. 🤷

File: synapse/storage/databases/main/receipts.py [hunk 1]
```diff
@@ -843,8 +830,8 @@ def _insert_graph_receipt_txn(
         else:
             keyvalues["thread_id"] = thread_id
 
-        self.db_pool.simple_upsert_txn(
-            txn,
+        await self.db_pool.simple_upsert(
```
Comment (clokep) [inline synapse/storage/databases/main/receipts.py:833]: I think you still want to use the `_txn` version of this and keep everything inside the same transaction?

Comment (clokep) [reply synapse/storage/databases/main/receipts.py:833]: I thought we inserted both the linearized and graph receipts in the same transaction?

**Comment by realtyem** [comment]
Should I separate the cache invalidation to move the 'linearized' related up near it's own transaction instead of hiding it inside another unrelated function?

**Review by erikjohnston — APPROVED** [review]
Thanks!

**Pull request merged**